### PR TITLE
[Cpp Graph] Update Falcon HF para and support Falcon-180B

### PR DESCRIPTION
**Pull request by zhentaoyu**
## Type of Change

bug fix, no api change

## Description

detail description 
JIRA ticket: None
falcon model updates some parameter names, see [here](https://huggingface.co/tiiuae/falcon-40b/blob/main/config.json)
```json
{
  "alibi": false,
  "apply_residual_connection_post_layernorm": false,
  "architectures": [
    "FalconForCausalLM"
  ],
  "attention_dropout": 0.0,
  "auto_map": {
    "AutoConfig": "configuration_falcon.FalconConfig",
    "AutoModel": "modeling_falcon.FalconModel",
    "AutoModelForSequenceClassification": "modeling_falcon.FalconForSequenceClassification",
    "AutoModelForTokenClassification": "modeling_falcon.FalconForTokenClassification",
    "AutoModelForQuestionAnswering": "modeling_falcon.FalconForQuestionAnswering",
    "AutoModelForCausalLM": "modeling_falcon.FalconForCausalLM"
  },
  "bias": false,
  "bos_token_id": 11,
  "eos_token_id": 11,
  "hidden_dropout": 0.0,
  "hidden_size": 8192,
  "initializer_range": 0.02,
  "layer_norm_epsilon": 1e-05,
  "model_type": "falcon",
  "new_decoder_architecture": true,
  "num_attention_heads": 128,
  "num_hidden_layers": 60,
  "num_kv_heads": 8,
  "parallel_attn": true,
  "torch_dtype": "bfloat16",
  "transformers_version": "4.27.4",
  "use_cache": true,
  "vocab_size": 65024
}
```

## Expected Behavior & Potential Risk

Ext test

## How has this PR been tested?

Ext test

## Dependency Change?

None

### CODE REVIEW

**Comment by zhentaoyu** [comment]
https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/102/

**Comment by zhentaoyu** [comment]
`falcon-180B` logits comparison: (`prompt="What is the meaning of life?"`)

`torch-fp32`:
```bash
inputs {'input_ids': tensor([[1562,  304,  248, 4113,  275, 1063,   42]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1]])}
(Pdb) out['logits'].shape
torch.Size([1, 7, 65024])
(Pdb) out['logits'][0][-1][:32]
tensor([ -6.4001, -13.1985, -16.0812, -14.1584,   1.1085,  -5.9462,  -7.0403,
         -1.1525, -17.9086, -17.9111, -17.9089,   0.2117,  -4.4417,  -2.3658,
         -2.7325,  -4.7559,  -3.8219,  -3.7943,  -3.2608,  -3.4782,  -5.8313,
         -0.0970,  -2.8514,  -2.2135,   0.6608,  -2.6231,   0.9718,  -1.4672,
         -0.7133,  -1.2605,  -1.4451,  -1.0655], grad_fn=<SliceBackward0>)
```

`fp32 ggml`:
```bash
What is the meaning of life?input_ids: 1562,304,248,4113,275,1063,42,

 logits:
   -6.417377,  -13.115252,  -16.080769,  -14.130390,    1.074707,   -5.805538,   -7.018699,  
   -1.167375,  -17.928669,  -17.931366,   -17.929062,    0.211842,   -4.451574,   -2.365051, 
    -2.738035,   -4.743909,   -3.841066,   -3.787909,   -3.260063,   -3.465620,   -5.804641,   
    -0.125300,   -2.884983,   -2.214450,    0.686065,   -2.610334,    0.957098,   -1.478186, 
   -0.706896,   -1.274001,   -1.438719,   -1.083174,
```

`q4_0 ggml`:
```bash
What is the meaning of life?input_ids: 1562,304,248,4113,275,1063,42,

 logits:
   -6.237789,  -12.688745,  -16.203352,  -14.168602,    1.278341,   -5.052035,   -6.667830,  
  -1.069596,  -17.688431,  -17.702812,  -17.708504,    0.136256,   -4.537621,   -1.995419,  
  -2.781709,   -4.412484,   -3.837585,   -3.449815,   -2.874452,   -3.493387,   -5.665998,  
  0.203061,   -2.141753,   -1.903326,    0.739224,   -2.560589,    0.600809,   -1.259985,  
  -0.576035,   -1.520468,   -1.061628,   -1.033232,
```

**Comment by zhentaoyu** [comment]
> https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/147/

pass ext test, `q4-j-b128` has ass issue because `falcon-7b`'s `n_embd=4544` which is not divisible by 128

**Comment by zhentaoyu** [comment]
perf reference:
command: `numactl -l -C 56-111 ./build/bin/run_falcon -m q4.bin -t 56 -c 2048 -b 2000 -n 32  --seed 16 -p`

| input tokens |  first token (avg, ms) | next token (avg, ms) |
|---------------|------------------------|-----------------------|
|                                        Q4_0                                           |
|   32               |  512.36                       |          700.90            |
|   512             |    468.95                     |          737.04           |
| 1024             |     601.17                    |            889.52         |
|                                      Q4_J_B128                                    |
| 32                 |          37.42                |            691.95          |
|  512              |          37.98                |             682.69         |
| 1024             |          53.24                |             707.83         |

File: intel_extension_for_transformers/llm/runtime/graph/models/falcon/falcon.h [hunk 1]
```diff
@@ -30,6 +30,8 @@ static const model_scratch falcon_mem_req(int n_layers) {
     // TODO(hengyu): add more variants besides 6B
     case 60:
       return {2 * 2048ull * MB, 2 * 2048ull * MB, 2 * 4096ull * MB};
+    case 80:
+      return {3 * 2048ull * MB, 3 * 2048ull * MB, 3 * 4096ull * MB};
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/models/falcon/falcon.h:34]: please make sure the size is fit not too large 
since  layer 60 is  2 * 2048, maybe layer 80 should be 2.7 ?

File: intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py [hunk 2]
```diff
@@ -67,14 +67,18 @@ def main(args_in: Optional[List[str]] = None) -> None:
...
+        print("Model architecture not supported: " + hparams["architectures"][0])
+        sys.exit(1)
     print("Loading model: ", dir_model)
     model = AutoModelForCausalLM.from_pretrained(dir_model, config=config, torch_dtype=torch.float16
                     if ftype == 1 else torch.float32, low_cpu_mem_usage=True, trust_remote_code=True)
     print("Model loaded: ", dir_model)
 
-    n_head_kv = hparams.get("n_head_kv", 1)
-    n_head = hparams["n_head"]
+    n_head_kv = hparams.get("num_kv_heads", 1)
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py:80]: why change n_head_kv to num_kv_heads

**Review by a32543254 — APPROVED** [review]
LGTM

File: intel_extension_for_transformers/llm/runtime/graph/models/falcon/falcon.h (hunk 1 above)
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/models/falcon/falcon.h:34]: layer 60 is for 40B, layer 80 is for 180 B. Width (param) and depth (layer) both need to be considered. The mem scratch way should be upgraded into auto-alloc arch in the future.

File: intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py (hunk 2 above)
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py:80]: HF `config.json` updated

**Review by intellinjun — APPROVED** [review]
LGTM

**Review by zhenwei-intel — APPROVED** [review]

**Review by airMeng — APPROVED** [review]
